services/wallet/activity/service.go

```diff
@@ -0,0 +1,123 @@
+package activity
+
+import (
+	"context"
+	"database/sql"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/event"
+	"github.com/ethereum/go-ethereum/log"
+
+	w_common "github.com/status-im/status-go/services/wallet/common"
+	"github.com/status-im/status-go/services/wallet/walletevent"
+)
+
+const (
+	// FilterResponse json is sent as a message in the EventActivityFilteringDone event
+	EventActivityFilteringDone walletevent.EventType = "wallet-activity-filtering-done"
+)
+
+type Service struct {
+	db        *sql.DB
+	eventFeed *event.Feed
+
+	context  context.Context
+	cancelFn context.CancelFunc
+	wg       sync.WaitGroup
+	mu       sync.Mutex
+}
+
+func NewService(db *sql.DB, eventFeed *event.Feed) *Service {
+	return &Service{
+		db:        db,
+		eventFeed: eventFeed,
+	}
+}
+
+type ErrorCode = int
+
+const (
+	ErrorCodeSuccess ErrorCode = iota + 1
+	ErrorCodeFilterCanceled
+	ErrorCodeFilterFailed
+	ErrorCodeServiceFailed
+)
+
+type FilterResponse struct {
+	Activities       []Entry   `json:"activities"`
+	ThereMightBeMore bool      `json:"thereMightBeMore"`
+	ErrorCode        ErrorCode `json:"errorCode"`
+}
+
+// FilterActivityAsync allows only one filter task to run at a time
+// and it cancels the current one if a new one is started
+// All calls will trigger an EventActivityFilteringDone event with the result of the filtering
+func (s *Service) FilterActivityAsync(ctx context.Context, addresses []common.Address, chainIDs []w_common.ChainID, filter Filter, offset int, limit int) error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	// If a previous task is running, cancel it and wait to finish
+	if s.cancelFn != nil {
+		s.cancelFn()
+		s.wg.Wait()
+	}
+
+	if ctx.Err() == nil {
```

Comment:
Minor thing: Arrow anti-pattern can be prevented by simply 
```suggestion
	if ctx.Err() != nil {
	    return nil
	}
```